pkg/kubelet/eviction/memory_threshold_notifier_test.go

```diff
@@ -0,0 +1,305 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eviction
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/clock"
+	statsapi "k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1"
+	evictionapi "k8s.io/kubernetes/pkg/kubelet/eviction/api"
+)
+
+const testCgroupPath = "/sys/fs/cgroups/memory"
+
+func nodeSummary(available, workingSet, usage resource.Quantity, allocatable bool) *statsapi.Summary {
+	availableBytes := uint64(available.Value())
+	workingSetBytes := uint64(workingSet.Value())
+	usageBytes := uint64(usage.Value())
+	memoryStats := statsapi.MemoryStats{
+		AvailableBytes:  &availableBytes,
+		WorkingSetBytes: &workingSetBytes,
+		UsageBytes:      &usageBytes,
+	}
+	if allocatable {
+		return &statsapi.Summary{
+			Node: statsapi.NodeStats{
+				SystemContainers: []statsapi.ContainerStats{
+					{
+						Name:   statsapi.SystemContainerPods,
+						Memory: &memoryStats,
+					},
+				},
+			},
+		}
+	}
+	return &statsapi.Summary{
+		Node: statsapi.NodeStats{
+			Memory: &memoryStats,
+		},
+	}
+}
+
+func newTestMemoryThresholdNotifier(threshold evictionapi.Threshold, cgpath string, factory NotifierFactory, clock clock.Clock) (*memoryThresholdNotifier, error) {
```

Comment:
we need the reference to the original factory and clock so that we can call methods on them that are not in the interface, e.g. clock.Step or notifierFactory.On.
But we can assume we are using the testCgroupPath.